man/en/tp-magic-config.1
.\" tp-magic-config - 2007.08.07
.TH TP-MAGIC-CONFIG 1 "07 August 2007" "2007.08.07" "tp-magic-config"
.SH NAME
tp-magic-config -- Helps creating 'Magic' tool plugins for Marx Paint(1)

.SH SYNOPSYS
.TP 16
.B tp-magic-config [\-\-apiversion | \-\-version | \-\-cflags | \-\-pluginprefix | \-\-plugindocprefix | \-\-dataprefix | \-\-localpluginprefix | \-\-localdataprefix]

.SH DESCRIPTION
\fItp-magic-config\fP is a simple shell script that responds with various
pieces of information about the currently-installed version of
\fIMarx Paint\fP(1) that are useful when building 'Magic' tool plugins.

.SH OPTIONS
.TP 8
.B \-\-apiversion
Outputs the version of the \fIMarx Paint\fP 'Magic' tool plugin API that the
installed copy of \fIMarx Paint\fP supports. (For API compatibility testing.)
.TP 8
.B \-\-version
Outputs the version of \fIMarx Paint\fP that \fItp-magic-config\fP
corresponds to.
.TP 8
.B \-\-cflags
Outputs the compiler flags that \fIMarx Paint\fP 'Magic' tool plugins should
be compiled with. (For example, a "\-I" include path option that tells the
compiler where it can find the plugin API header file, "tp_magic_config.h",
that plugins must #include.)
.TP 8
.B \-\-pluginprefix
Outputs the system directory where the installed copy of \fIMarx Paint\fP expects
to find 'Magic' tool plugins (".so" shared objects).
(e.g., "/usr/share/marxpaint/plugins")
.TP 8
.B \-\-localpluginprefix
Outputs the user directory where the installed copy of \fIMarx Paint\fP expects
to find 'Magic' tool plugins (".so" shared objects).
(e.g., "/home/username/.marxpaint/plugins")
.TP 8
.B \-\-plugindocprefix
Outputs the directory where the installed copy of \fIMarx Paint\fP expects
to find documentation for 'Magic' tool plugins (".html" and ".txt" files).
\fIMarx Paint's\fP main documentation includes a link to this directory
under the section on "Magic" tools.
.TP 8
.B \-\-dataprefix
Outputs the system directory where the installed copy of \fIMarx Paint\fP keeps
its global data files (e.g., "/usr/share/marxpaint/"). This is the same value
that plugins installed system-wide will receive in the "data_directory" string
within the "magic_api" structure sent to the plugins' functions.
.TP 8
.B \-\-localdataprefix
Outputs the user directory where the installed copy of \fIMarx Paint\fP
expects plugins to install their local data files.
(e.g., "/home/username/.marxpaint/plugins/data"). This is the same value
that plugins installed locally will receive in the "data_directory" string
within the "magic_api" structure sent to the plugins' functions.

.SH SYSTEM-WIDE SHELL EXAMPLES
$ gcc -shared `tp-magic-config --cflags` my_plugin.c -o my_plugin.so
.br
# cp my_plugin.so `tp-magic-config \-\-pluginprefix`
.br
# cp my_plugin_icon.png `tp-magic-config \-\-dataprefix`/images/magic
.br
# cp my_plugin.html `tp-magic-config \-\-plugindocrefix`/html
.br
# cp my_plugin.txt `tp-magic-config \-\-plugindocrefix`

.SH LOCAL SHELL EXAMPLES
$ gcc -shared `tp-magic-config --cflags` my_plugin.c -o my_plugin.so
.br
$ mkdir -p `tp-magic-config \-\-localpluginprefix`
.br
$ cp my_plugin.so `tp-magic-config \-\-localpluginprefix`
.br
$ mkdir -p `tp-magic-config \-\-localdataprefix`/images/magic
.br
$ cp my_plugin_icon.png `tp-magic-config \-\-localdataprefix`/images/magic

.SH SYSTEM-WIDE MAKEFILE EXAMPLE
MAGIC_CFLAGS=$(shell tp-magic-config --cflags)
.br
MAGIC_PREFIX=$(shell tp-magic-config --pluginprefix)
.br
MAGIC_DOC_PREFIX=$(shell tp-magic-config --plugindocprefix)
.br
DATA_PREFIX=$(shell tp-magic-config --dataprefix)
.PP
all: my_plugin.so
.PP
my_plugin.so: my_plugin.c
.PP	$(CC) -shared $(MAGIC_CFLAGS) my_plugin.c -o my_plugin.so
.PP
install: install-so install-data install-docs
.PP
install-so:
.br
	mkdir -p $(MAGIC_PREFIX)
.br
	cp my_plugin.so $(MAGIC_PREFIX)/
.br
	chmod 644 $(MAGIC_PREFIX)/my_plugin.so
.PP
install-data:
.br
	mkdir -p $(DATA_PREFIX)
.br
	cp icons/my_plugin_icon.png $(DATA_PREFIX)/images/magic/
.br
	chmod 644 $(DATA_PREFIX)/images/magic/my_plugin_icon.png
.PP
install-docs:
.br
	mkdir -p $(MAGIC_DOC_PREFIX)
.br
	cp docs/my_plugin.html $(MAGIC_DOC_PREFIX)/html/
.br
	chmod 644 $(MAGIC_DOC_PREFIX)/html/my_plugin.html
.br
	cp docs/my_plugin.txt $(MAGIC_DOC_PREFIX)/
.br
	chmod 644 $(MAGIC_DOC_PREFIX)/my_plugin.txt

.SH AUTHOR
Bill Kendrick.  <bill@newbreedsoftware.com>


.SH "SEE ALSO"
.BR marxpaint (1),
.PP
And documentation within /usr/[local/]share/doc/marxpaint/.



docs/CHANGES.txt
CHANGES.txt for Marx Paint

Marx Paint - A simple drawing program for communists.

Copyright (c) 2002-2021
Various contributors (see below, and AUTHORS.txt)
http://www.marxpaint.org/

$Id$

2021.August.15 (0.9.27)
  * Documentation updates
    ---------------------
    * Ensured Marx Paint's built-in help ("marxpaint --help"),
      Unix manual ("man page"), and command-line option
      tab-completion list, and "OPTIONS" documentation all cover
      the full set of Marx Paint options.

    * Reorganized the "Available Options" section of "OPTIONS"
      documentation (to match the organization found in the UI
      of "Marx Paint Config.")

    * Reorganized the options listed in Marx Paint's man page,
      to more closely (but not precisely) match the organization
      noted above.

    * marxpaint(1) manpage is now managed & translatable via the
      'marxpaint-docs' project, just like the HTML & plaintext
      documentation files (since 0.9.26).

2021.June.28 (0.9.26)
  * New Features
    ------------
    * Larger UI buttons
      Override the ages-old default 48x48 size for UI buttons
      (via a new "buttonsize" option).  Note: If the button size
      would be too large for the required parts of the UI to appear
      within Marx Paint's window size, the largest possible size will
      be used.
      Pere Pujal i Carabantes <pere@fornol.no-ip.org>

    * It's now possible to increase how many rows of buttons are shown
      for the color palette at the bottom of the screen.
      (Use the "colorsrows" option.)
      Pere Pujal i Carabantes <pere@fornol.no-ip.org>

    * Adding sub-tools to the Fill tool:
      + Solid -- The classic fill tool; click to fill 
      + Linear -- A linear gradient; click and drag to adjust,
        release mouse to finish
      + Radial -- A radial gradient; click to fill
      (The gradients transition from the current color to the
      background that's being filled.)

  * New Magic Tools:
    ----------------
    * Checkerboard
      Cover the entire canvas with a checkboard pattern using the
      current color.  Drag to change the size of the squares.

    * Clone
      Clones (copies, via painting) part of a drawing to another position.

    * Pixels
      Draws large squares, for classic computer 'pixel art' style drawings.

  * Documentation updates
    ---------------------
    * Documentation is now maintained in a separate source repository,
      'marxpaint-docs', which allows for easier and more consisten
      translation of the documentation.  Pulling in initial version of
      these docs, which contain a variety of minor updates (and also a
      few which had previously only been available as plain TXT are now
      available as HTML).

      + French translation
        Jacques Chion <jacques.chion@orange.fr>

      + Galician
        Miguel Anxo Bouzada <mbouzada@gmail.com>

      + Japanese
        TOYAMA Shin-ichi <shin1@wmail.plala.or.jp>

    * Adding documentation that covers some of the
      environment variables that Marx Paint utilizes.

    * Expanded the steps for making and posting releases.

    * Documented how to build Marx Paint using old macOS
      Mark K. Kim <markuskimius@gmail.com>

    * Added more credits to manpage.  Avoiding word-wrapping
      URLs and email addresses.  Saved out as UTF-8, rather than
      Latin-1 encoding.

  * Other Improvements
    ------------------
    * Fill tools do a better job at filling around
      antialiased edges, and apply some blending.

    * Reduce CPU usage by increasing delay in main loop
      from 1ms to 10ms, and only using SDL's Timer subsystem
      when scrolling happens.
      (Details at
      https://sourceforge.net/p/marxpaint/marxpaint/ci/7727b995c53df208596eff89ac1acb954a16098c/)
      Mark K. Kim <markuskimius@gmail.com>
      (h/t @bbugwong on Twitter for reporting the issue)

    * If parent of export directory doesn't exist, Marx Paint will
      try to create it as well.
      (Only one level up; e.g., with an export location like
      /path/to/Pictures/MarxPaint/, it will try to create
      /path/to/Pictures/ if it doesn't exist, but not its parents).
      h/t Tim Dickson

  * Translation Updates:
    --------------------
    * Albanian
      Besnik Bleta <besnik@programeshqip.org>

    * Catalan
      Pere Pujal i Carabantes <pere@fornol.no-ip.org>

    * Galician
      Miguel Bouzada <mbouzada@gmail.com>

    * Icelandic
      Sveinn í Felli <sv1@fellsnet.is>

    * Occitan
      Quentin PAGÈS <quentinantonin@free.fr>

  * Magic Tool Improvememnts:
    -------------------------
    * A new "xorpixel()" function has been added to the Magic Tools API.

  * Bug Fixes:
    ----------
    * Halftone magic tool wasn't loading due to a problem
      with its sound effect file.

  * Ports & Building
    ----------------
    * [WIP] Apple iOS port
      Mark K. Kim <markuskimius@gmail.com>

2020.December.27 (0.9.25)
  * New Features
    ------------
    * Export drawings:
      * GIF export option from the Open -> Slideshow dialog.
        (Closes https://sourceforge.net/p/marxpaint/feature-requests/191/)
        * Utilizes "gifenc", public domain by Marcel Rodrigues:
          https://github.com/lecram/gifenc
        * Depends on "libimagequant", from the "pngquant2" project:
          https://github.com/ImageOptim/libimagequant
          (GPL v3 or later, for Free/Libre Open Source Software)
          (h/t TOYAMA Shin-ichi for adding support for versions < 2.8.0)
      * Single image (PNG) export from the Open dialog.
        (Closes https://sourceforge.net/p/marxpaint/feature-requests/192/)
      * Both utilize XDG user dirs settings, if available, to determine
        where to store them, or "~/Pictures", otherwise.  In those cases,
        files are stored in a new "MarxPaint" subdirectory.
      * The "--exportdir" option can be used to override the default
        location (but a "MarxPaint" subdir. will NOT be placed there).


src/tp-magic-config.sh
#!/bin/sh

# tp-magic-config

# "Tux Paint Magic Config"
# Tool that reports compiler options used when buidling Magic Tool
# shared objects for Tux Paint

# (c) Copyright 2007, by Bill Kendrick
# bill@newbreedsoftware.com
# http://www.tuxpaint.org/

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
# (See COPYING.txt)

# Note: "__VERSION__" and "__INCLUDE__" are replaced by values
# in Tux Paint's Makefile, via 'sed', by the 'make install-magic-plugin-dev'
# target.

# July 5, 2007 - July 8, 2007


if [ $# -ne 0 ]; then
  if [ $1 = "--version" ]; then
    echo "__VERSION__"
    exit
  fi
  if [ $1 = "--apiversion" ]; then
    echo "__APIVERSION__"
    exit
  fi
  if [ $1 = "--cflags" ]; then
    echo `sdl-config --cflags` -I__INCLUDE__
    exit
  fi
fi

echo "Usage: tp-magic-config [--apiversion] [--version] [--cflags]"

